src/coreclr/jit/compiler.cpp

```diff
@@ -9735,7 +9735,10 @@ void Compiler::gtChangeOperToNullCheck(GenTree* tree, BasicBlock* block)
 {
     assert(tree->OperIs(GT_FIELD, GT_IND, GT_OBJ, GT_BLK));
     tree->ChangeOper(GT_NULLCHECK);
-    tree->ChangeType(TYP_INT);
+    if (!varTypeIsIntegralOrI(tree))
+    {
+        tree->ChangeType(TYP_INT);
+    }
```

Comment:
This seems unsatisfying; I have the same question as @jakobbotsch elsewhere: what if the tree is a 1-byte struct? Couldn't we still overrun a page?